File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        //basically we are going to take courses which has indegree 0
        vector<int>indegree(numCourses);
        vector<vector<int>>adj(numCourses);
        for(auto p: prerequisites){
            adj[p[1]].push_back(p[0]);
            indegree[p[0]]++;
        }
        queue<int>queue;
        for(int i=0;i<numCourses;i++){
            if(indegree[i]==0){
                queue.push(i);
            }
        }
        int coursesTaken=0;
        while(!queue.empty()){
            coursesTaken++;
            int node = queue.front();
            queue.pop();
            for(int neigh:adj[node]){
                indegree[neigh]--;
                if(indegree[neigh]==0){
                    queue.push(neigh);
                }
            }
        }
        return coursesTaken==numCourses;
    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        
        List<List<Integer>>adjList = new ArrayList<>(numCourses);
        for(int i=0;i<numCourses;i++){
            adjList.add(new ArrayList<>());
        }
        int[] inDegree = new int[numCourses];
        for(int[] prereq: prerequisites){
            adjList.get(prereq[1]).add(prereq[0]);
            inDegree[prereq[0]]++;
        }
        Queue<Integer>queue = new LinkedList<>();
        for(int i=0;i<numCourses;i++){
            if(inDegree[i]==0){
                queue.add(i);
            }
        }
        int completed=0;
        while(!queue.isEmpty()){
            int c = queue.poll();
            if(completed==numCourses) return true;
            for(int course:adjList.get(c)){
                inDegree[course]--;
                if(inDegree[course]==0){
                    queue.add(course);
                }
            }
            completed++;
        }
        return completed==numCourses;
    }
}